From stable-bounces@linux.kernel.org  Wed Jun  1 03:47:52 2005
Date: Wed, 1 Jun 2005 12:47:30 +0200
From: Andi Kleen <ak@suse.de>
To: stable@kernel.org
Subject: [PATCH] x86_64: Fix ptrace boundary check

Don't allow accesses below register frame in ptrace

There was a "off by one quad word" error in there.

Found and fixed by John Blackwood

Signed-off-by: Andi Kleen <ak@suse.de>
Signed-off-by: Chris Wright <chrisw@osdl.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 arch/x86_64/kernel/ptrace.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

--- linux-2.6.11.10.orig/arch/x86_64/kernel/ptrace.c	2005-05-16 10:50:31.000000000 -0700
+++ linux-2.6.11.10/arch/x86_64/kernel/ptrace.c	2005-06-08 22:40:26.000000000 -0700
@@ -247,7 +247,7 @@
 			break;
 
 		switch (addr) { 
-		case 0 ... sizeof(struct user_regs_struct):
+		case 0 ... sizeof(struct user_regs_struct) - sizeof(long):
 			tmp = getreg(child, addr);
 			break;
 		case offsetof(struct user, u_debugreg[0]):
@@ -292,7 +292,7 @@
 			break;
 
 		switch (addr) { 
-		case 0 ... sizeof(struct user_regs_struct): 
+		case 0 ... sizeof(struct user_regs_struct) - sizeof(long):
 			ret = putreg(child, addr, data);
 			break;
 		/* Disallows to set a breakpoint into the vsyscall */
